elfo-network/src/frame/read.rs

```diff
@@ -0,0 +1,235 @@
+                            position: 0,
+                        };
+                        (bytes_consumed, 0)
+                    }
+                }
+            }
+            LZ4DecodingState::MessageParsing {
+                compressed_size,
+                position,
+            } => (compressed_size, position),
+        };
+
+        let decompressed_buffer = self.decompressed_buffer.get_ref();
+        if position >= decompressed_buffer.len() {
+            self.compressed_buffer.consume(compressed_size);
+            self.state = LZ4DecodingState::FrameDecompression;
+            return Ok(FramedReadState::Done { decoded: None });
+        }
+
+        let envelope_buffer = &decompressed_buffer[position..];
+        let codec_state = codec::decode::decode(envelope_buffer, &mut self.stats.decode_stats)?;
+        match codec_state {
+            DecodeState::NeedMoreData { .. } => {
+                Err(eyre!("lz4 decompressed data contains truncated envelopes"))
+            }
+            DecodeState::Done {
+                bytes_consumed,
+                decoded,
+            } => {
+                self.state = LZ4DecodingState::MessageParsing {
+                    compressed_size,
+                    position: position + bytes_consumed,
+                };
+                Ok(FramedReadState::Done {
+                    decoded: Some(decoded),
+                })
+            }
+        }
+    }
+
+    fn mark_filled(&mut self, count: usize) {
+        self.compressed_buffer.mark_filled(count);
+    }
+
+    fn take_stats(&mut self) -> FramedReadStats {
+        std::mem::take(&mut self.stats)
+    }
+}
+
+pub(crate) struct NoneFramedRead {
+    buffer: ReadBuffer,
+    stats: FramedReadStats,
+}
+
+impl NoneFramedRead {
+    pub(crate) fn new() -> Self {
+        Self {
+            buffer: ReadBuffer::with_capacity(DECOMPRESSED_DATA_BUFFER_CAPACITY),
+            stats: Default::default(),
+        }
+    }
+}
+
+impl FramedReadStrategy for NoneFramedRead {
+    fn read(&mut self) -> Result<FramedReadState<'_>> {
+        let codec_state =
+            codec::decode::decode(self.buffer.as_slice(), &mut self.stats.decode_stats)?;
+        match codec_state {
+            DecodeState::NeedMoreData {
+                total_length_estimate,
+            } => {
+                self.buffer.extend_to_contain(total_length_estimate);
+                let min_bytes = total_length_estimate - self.buffer.len();
+                Ok(FramedReadState::NeedMoreData {
+                    buffer: self.buffer.remaining(),
+                    min_bytes,
+                })
+            }
+            DecodeState::Done {
+                bytes_consumed,
+                decoded,
+            } => {
+                self.stats.decompress_stats.total_uncompressed_bytes += bytes_consumed as u64;
+                self.buffer.consume(bytes_consumed);
+                Ok(FramedReadState::Done {
+                    decoded: Some(decoded),
```

Comment:
It's suspicious that there is no `None` variant, it indicates for me not ideal LZ4 implementation.